packages/flutter/lib/src/material/flat_button.dart

```diff
@@ -336,3 +156,56 @@ class FlatButton extends StatelessWidget {
     properties.add(DiagnosticsProperty<MaterialTapTargetSize>('materialTapTargetSize', materialTapTargetSize, defaultValue: null));
   }
 }
+
+/// The type of of FlatButtons created with [FlatButton.icon].
+///
+/// This class only exists to give FlatButtons created with [FlatButton.icon]
+/// a distinct class for the sake of [ButtonTheme]. It can not be instantiated.
+class FlatButtonWithIcon extends FlatButton {
```

Comment:
I have an idea to both reduce the number of `is` checks for the isIcon case and allow users to correctly extend an IconTheme.

Define an empty mixin called `MaterialIconButton`, or similar.

```
/// Something something indicates that we use icon theme.s
mixin MaterialIconButton {}
```

Then make `FlatButtonWithIcon` and friends mix this in instead of being public class/private constructor.

```
_FlatButtonWithIcon extends FlatButton with MaterialIconButton { ... }
```

You can then simplify the isIcon check to a single `is`:
```
 bool _isIconButton(MaterialButton button)  => button is MaterialIconButton
```

And users can mixin this class to make custom subclasses take the icon config as a bonus.

Reply:
Good idea. I added MaterialButtonWithIconMixin and made the FooButtonWithIcon classes private.